### ISSUE TITLE
as_ptr()'s use of *mut is misleading

### ISSUE THREAD

**Issue by kornelski** [sfackler/foreign-types#24]
`as_ptr()` returns a `*mut _` pointer. This isn't an error by itself, but under Rust's memory model, this pointer is derived from shared/immutable `&self`, and therefore doesn't have permission to mutate `self`, so the `mut` of the pointer is misleading.

https://github.com/sfackler/foreign-types/blob/393f6ab5a5dc66b8a8e2d6d880b1ff80b6a7edc2/foreign-types-shared/src/lib.rs#L89

I assume this is for convenience, because C APIs aren't diligent about `const` vs `mut` distinction. However, this is a gotcha, because C APIs that take `*mut` may actually mutate the object, and that is UB from Rust's perspective. 

Could you add `.as_mut_ptr()` that takes `&mut self` to provide a pointer safe for mutation?

**Comment by sfackler** [comment]
The structure of `Opaque` is designed to prevent mutation through the `*mut` from being UB: https://github.com/sfackler/foreign-types/blob/393f6ab5a5dc66b8a8e2d6d880b1ff80b6a7edc2/foreign-types-shared/src/lib.rs#L14.

Even for C APIs that are const-correct, C's notion of constness is distinct enough from Rust's that `& <-> *const` and `&mut <-> *mut` doesn't really work.

**Comment by icmccorm** [comment]
I've developed a version of Miri that can execute foreign functions by interpreting LLVM bytecode, and it found a Tree Borrows violation related to mutating through `self.as_ptr()`. Writing through the mutable pointer created from `self.as_ptr()` is considered undefined behavior under this model, since this pointer is given a `Frozen` access tag as a child tag of the implicit `&self` parameter to `as_ptr`.

Here's [a minimal example](https://github.com/icmccorm/lcms2-recreation) that can recreate the error with an unmodified version of Miri, without the FFI. In particular, this error is triggered when mutating through `self.as_ptr()` within a method that receives `&mut self`, where `self` is an instance of `ForeignTypeRef`.

```
impl ExampleRef {
    fn invalid_mutation(&mut self) {
        let raw_ptr: *mut ExampleInner = self.as_ptr();
        unsafe {
            (*raw_ptr).inner = 1;
        }
    }
}
```

Also, I was a bit unsure about what you meant here:

> C's notion of constness is distinct enough from Rust's that & <-> *const and &mut <-> *mut doesn't really work.

Are you saying that the compiler's treatment of `&T` differs enough from `* const` in C (even under a const-correct API) that this borrowing violation isn't actually problematic in practice?

**Comment by sfackler** [comment]
The `&self` is a pointer to UnsafeCell (semantically through a PhantomData). Why is it being tagged as Frozen?

**Comment by kornelski** [comment]
Could it be because `UnsafeCell` in `PhantomData` has zero size, and doesn't exist in memory? If it was `Opaque(UnsafeCell<ActualForeignSizedType>)` then it would map to an address space.

**Comment by madsmtm** [comment]
Yeah, I think there's a difference between `PhantomData<UnsafeCell<*mut ()>>` and `UnsafeCell<PhantomData<*mut ()>>`, with only the latter registering as using interior mutability

**Comment by madsmtm** [comment]
I'll note again: `rustc` _currently_ treats `UnsafeCell` differently whether it's inside `PhantomData` or not! See [this playground](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=eadb927c958434ba2094d88164a43af8), click "Show LLVM IR", and inspect the result; the function taking `&PhantomData<UnsafeCell<*mut ()>>` gets marked with `noalias readonly`, while `&UnsafeCell<PhantomData<*mut ()>>` gets marked with `readnone`.

If you [add extern types to the mix](https://play.rust-lang.org/?version=nightly&mode=release&edition=2021&gist=e5c9908ff28033ebcb1f20df5980b48f), you'll see the exact same LLVM IR as for `&UnsafeCell<PhantomData<*mut ()>>` (LLVM will even merge the functions because of that).

**Comment by madsmtm** [comment]
See also discussion from https://github.com/rust-lang/unsafe-code-guidelines/issues/236#issuecomment-1637072799, this is still not determined

**Comment by icmccorm** [comment]
> If you [add extern types to the mix](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=eadb927c958434ba2094d88164a43af8), you'll see the exact same LLVM IR

@madsmtm I think you accidentally used the same link twice.

Based on [this pull request](https://github.com/rust-lang/rust/pull/114559#issuecomment-1677792712), allowing `PhantomData` to apply interior mutability would be a breaking change, so it's not likely to be resolved in this direction soon. It seems like it would be worthwhile to change `Opaque` for now until the semantics are settled.

**Comment by madsmtm** [comment]
> I think you accidentally used the same link twice.

Apologies, I've fixed it now.

**Comment by taikulawo** [comment]
From my view, `Opaque` is zero size type, 
So SslRef should be ZST 
```
pub struct SslRef(foreign_types::Opaque);
```
Does we convert pointer(for example, `*mut ffi::SSL`) to SslRef is soundness?

```rs
#[inline]
unsafe fn from_ptr_mut<'a>(ptr: *mut Self::CType) -> &'a mut Self {
    &mut *(ptr as *mut _)
}
```
